File: day1/ruby/solution.rb
# frozen_string_literal: true

class DayOne2023
  def initialize
    @lines = File.read("../input1.txt").split("\n")
    puts part_one
    puts part_two
  end

  def part_one
    @lines.sum do |line|
      digits = line.scan(/\d/)
      first = digits.first
      last = digits.last
      "#{first}#{last}".to_i
    end
  end

  def part_two
    numbers = %w[one two three four five six seven eight nine]
    @lines.sum do |line|
      digits = line.scan(/(?=(one|two|three|four|five|six|seven|eight|nine|\d))/).flatten
      first = numbers.index(digits.first)&.next || digits.first
      last = numbers.index(digits.last)&.next || digits.last
      "#{first}#{last}".to_i
    end
  end
end

DayOne2023.new

File: day1/js/solution.js
const fs = require('fs');

class DayOne2023 {
  constructor() {
    this.lines = fs.readFileSync('../input1.txt', 'utf8').split('\n');
    console.log(this.partOne());
    console.log(this.partTwo());
  }

  convertWordToNumber(word) {
    const numbers = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'];
    return numbers.indexOf(word) + 1; // Returns 0 if word is not found
  }

  partOne() {
    return this.lines.reduce((sum, line) => {
      const digits = line.match(/\d/g) || [];
      const first = parseInt(digits[0], 10);
      const last = parseInt(digits[digits.length - 1], 10);
      if (!isNaN(first) && !isNaN(last)) {
        return sum + parseInt(`${first}${last}`, 10);
      }
      return sum;
    }, 0);
  }

  partTwo() {
    return this.lines.reduce((sum, line) => {
      const regex = new RegExp('one|two|three|four|five|six|seven|eight|nine|\\d', 'g');
      const digitWords = line.match(regex) || [];
      let first = digitWords[0];
      let last = digitWords[digitWords.length - 1];

      first = isNaN(first) ? this.convertWordToNumber(first) : parseInt(first, 10);
      last = isNaN(last) ? this.convertWordToNumber(last) : parseInt(last, 10);

      if (!isNaN(first) && !isNaN(last)) {
        return sum + parseInt(`${first}${last}`, 10);
      }
      return sum;
    }, 0);
  }
}

new DayOne2023();
